### ISSUE TITLE
Fix MLP optimiser divergence in the v1.6.1 release

### ISSUE THREAD

**Issue by shawcharles** [sheep-farm/Greeners#12]
## Reproduction

Hayashi now pins its numerical dependency to Greeners `v1.6.1` at `5568d757b234c813ac59872df5f975a246eb1c06`. Running Hayashi’s existing deterministic `mlp_simulated` case with that exact release returns:

```text
r_squared,-900762112543967200000000000000000000000000000000,NaN
```

The same data and Hayashi script pass against current Greeners `develop` (`1.6.2-dev`), reporting `R² = 0.9855345459601756` and agreeing with independent R `nnet` and Python scikit-learn references under the existing tolerance.

## Cause

In Greeners 1.6.1 `src/mlp.rs`, the backpropagation code defines `dy_hat = -err`, which is the loss gradient, then accumulates a momentum term and applies it with `w += v` / `b += v`. That moves parameters *along* the gradient and performs ascent. Current `develop` instead uses Adam and subtracts the gradient-scaled update.

## Requested change

- Retain the current public MLP API and make the optimiser a descent method.
- Add a deterministic regression test using a simple, learnable simulated regression and the normal Hayashi-like settings (`hidden=10`, `lr=0.01`, `epochs=200`). Assert finite fitted values and materially positive in-sample R-squared; avoid a brittle exact-weight assertion.
- Publish the correction in a tagged Greeners release.

## Downstream follow-up

After a tag is published, Hayashi can update its explicit Greeners tag/SHA and lockfile, then rerun the unchanged `mlp_simulated` case. This issue deliberately does not propose changing Hayashi’s MLP scripts, reference results, or tolerances.

**Comment by shawcharles** [comment]
Resolved by the tagged Greeners v1.6.2 release. The release contains the Adam descent update, and Hayashi dev now pins v1.6.2 at `ada1c6c828e04565d1acff3974d105ede6b937bc`. Hayashi CI and the full validation workflow are green against that baseline. Closing as completed.
